base/binaryplatforms.jl

```diff
@@ -0,0 +1,999 @@
+                    end
+                    # Otherwise, return the `string(VersionNumber(str))` version.
+                    return normver(v)
+                end
+                value = normver(value)
+            end
+
+            # I know we said only alphanumeric and dots, but let's be generous so that we can expand
+            # our support in the future while remaining as backwards-compatible as possible.  The
+            # only characters that are absolutely disallowed right now are `-`, `+`, ` ` and things
+            # that are illegal in filenames:
+            nonos = raw"""+- /<>:"'\|?*"""
+            if any(occursin(nono, tag) for nono in nonos)
+                throw(ArgumentError("Invalid character in tag name \"$(tag)\"!"))
+            end
+
+            # Normalize and reject nonos
+            value = lowercase(string(value))
+            if any(occursin(nono, value) for nono in nonos)
+                throw(ArgumentError("Invalid character in tag value \"$(value)\"!"))
+            end
+            tags[tag] = value
+        end
+
+        # Auto-map call_abi and libc where necessary:
+        if os == "linux" && !haskey(tags, "libc")
+            # Default to `glibc` on Linux
+            tags["libc"] = "glibc"
+        end
+        if os == "linux" && arch ∈ ("armv7l", "armv6l") && "call_abi" ∉ keys(tags)
+            # default `call_abi` to `eabihf` on 32-bit ARM
+            tags["call_abi"] = "eabihf"
+        end
+
+        # If the user is asking for strict validation, do so.
+        if validate_strict
+            validate_tags(tags)
+        end
+
+        # By default, we compare julia_version only against major and minor versions:
+        if haskey(tags, "julia_version") && !haskey(compare_strategies, "julia_version")
+            compare_strategies["julia_version"] = (a::String, b::String, a_comparator, b_comparator) -> begin
+                a = VersionNumber(a)
+                b = VersionNumber(b)
+                return a.major == b.major && a.minor == b.minor
+            end
+        end
+
+        return new(tags, compare_strategies)
+    end
+end
+
+# Other `Platform` types can override this (I'm looking at you, `AnyPlatform`)
+tags(p::Platform) = p.tags
+Base.haskey(p::AbstractPlatform, k::String) = haskey(tags(p), k)
+
+# Allow us to easily serialize Platform objects
+function Base.repr(p::Platform; context=nothing)
+    str = string(
+        "Platform(",
+        repr(arch(p)),
+        ", ",
+        repr(os(p)),
+        "; ",
+        join(("$(k) = $(repr(v))" for (k, v) in tags(p) if k ∉ ("arch", "os")), ", "),
+        ")",
+    )
+end
+
+# Simple equality definition; for compatibility testing, use `platforms_match()`
+Base.:(==)(a::AbstractPlatform, b::AbstractPlatform) = tags(a) == tags(b)
+
+function validate_tags(tags::Dict)
+    throw_invalid_key(k) = throw(ArgumentError("Key \"$(k)\" cannot have value \"$(tags[k])\""))
+    ARCHITECTURE_FLAGS = Dict(
```

Comment:
make this a const and move it outside?

Reply:
Is that preferable?  I thought it was better to have it embedded so that it doesn't show up in as many scopes.

Reply:
Otherwise, it will be created at every call, no? But maybe not performance sensitive so it doesn't matter? If you think about scope I think you would usually do it as


```
let
const FOO = ...
global function f(...)
    # use FOO
end
end
```

Reply:
Yeah, I kind of thought the compiler would notice it was constant at compile-time and would store the completed object.  It looks like it doesn't do that though.